### ISSUE TITLE
Web3 instantiation from subclass throws error

### ISSUE THREAD

**Issue by egeftx** [ethereum/web3.py#2586]
* Version: 5.30.0
* Python: 3.8
* OS: osx
* pip freeze output is intentionally omitted.

### What was wrong?
We use a sub-class of Web3, for the purpose of this example, call it SubclassWeb3, as follows:

```
class SubclassWeb3(Web3):
def __init__(self, ....):
  ... do bunch of stuff...
  super().__init__(*args, **kwargs)
  ... do bunch more stuff...
```

This setup worked fine with web3 version == 5.23.1. However, switching to web3==5.30.0 breaks this, with error

```
[rest of callstack omitted]
 File “[redacted]/web3/main.py”, line 251, in __init__
    self.attach_modules(modules)
  File “[redacted]/web3/main.py”, line 340, in attach_modules
    _attach_modules(self, modules)
  File “[redacted]/web3/_utils/module.py”, line 68, in attach_modules
    setattr(parent_module, module_name, module_class(w3))
  File “[redacted]/web3/module.py”, line 93, in __init__
    self.codec: ABICodec = web3.codec
AttributeError: ‘NoneType’ object has no attribute ‘codec’
```

This used to work in older versions because the parent_module is assumed to be a Web3 object and stored accordingly - [see logic](https://github.com/ethereum/web3.py/blob/2b229d2e16792e52bb47108e665c6a8937c188a2/web3/_utils/module.py#L33-L35).

Compare this to the somewhat brittle name checking that happens in 5.30.0 - [see logic](https://github.com/ethereum/web3.py/blob/ee5f0543ad91e4ba991997e35f50719ceb58135d/web3/_utils/module.py#L58-L60). This doesn't quite work because if this logic is being executed from a subclass of Web3, the name of the parent module appears as SubclassWeb3 (and not Web3).

### How can it be fixed?

Fill this section in if you know how this could or should be fixed.

**Comment by fselmo** [comment]
Hey @egeftx... yeah that check was definitely scrappy. We couldn't avoid a circular import there but importing ``web3`` locally seems to be the best way since we can then use ``isinstance()`` appropriately. I have a draft PR up. We should be able to resolve that in there (feel free to chime in over there if you see something) and scoop it into the next ``v5`` release as a bugfix soon. Thanks for catching that!

**Comment by fselmo** [comment]
Closed by #2587 (`v5`) and #2592 (`v6`)
